libnetwork/ipams/defaultipam/address_space.go

```diff
@@ -2,36 +2,64 @@ package defaultipam
 
 import (
 	"context"
-	"fmt"
 	"net/netip"
+	"slices"
 	"sync"
 
 	"github.com/containerd/log"
 	"github.com/docker/docker/libnetwork/internal/netiputil"
 	"github.com/docker/docker/libnetwork/ipamapi"
+	"github.com/docker/docker/libnetwork/ipamutils"
+	"github.com/docker/docker/libnetwork/ipbits"
 	"github.com/docker/docker/libnetwork/types"
 )
 
 // addrSpace contains the pool configurations for the address space
 type addrSpace struct {
-	// Master subnet pools, indexed by the value's stringified PoolData.Pool field.
+	// Ordered list of allocated subnets. This field is used for dynamic subnet
+	// allocations.
+	allocated []netip.Prefix
+	// Allocated subnets, indexed by their prefix. Values track address
+	// allocations.
 	subnets map[netip.Prefix]*PoolData
 
-	// Predefined pool for the address space
-	predefined           []netip.Prefix
-	predefinedStartIndex int
+	// predefined pools for the address space
+	predefined []*ipamutils.NetworkToSplit
 
 	mu sync.Mutex
 }
 
-func newAddrSpace(predefined []netip.Prefix) (*addrSpace, error) {
+func newAddrSpace(predefined []*ipamutils.NetworkToSplit) (*addrSpace, error) {
+	slices.SortFunc(predefined, func(a, b *ipamutils.NetworkToSplit) int {
+		return netiputil.PrefixCompare(a.Base, b.Base)
+	})
+
+	// We need to discard longer overlapping prefixes (sorted after the shorter
+	// one), otherwise the dynamic allocator might consider a predefined
+	// network is fully overlapped, go to the next one, which is a subnet of
+	// the previous, and allocate from it.
+	j := 0
+	for i := 1; i < len(predefined); i++ {
+		if predefined[j].Overlaps(predefined[i].Base) {
+			continue
+		}
+		j++
+		predefined[j] = predefined[i]
+	}
+
+	if len(predefined) > j {
+		j++
+	}
```

Comment:
Any nonzero pointers in the backing array between `predefined[j:len(predefined)]` are considered alive references for GC purposes, even after reslicing. The tail of the slice needs to be zeroed out to avoid a memory leak.

```suggestion
	}
	clear(predefined[j:len(predefined)])
```